new feature: refactor server-side PKI

mattp- opened the pull request:
### What does this PR do?
The end goal of this work is to support a non file/disk based interface for pub key management. Using the existing salt.cache interface allows  to leverage existing implementations of other storage backends (ie mysql, postgres, redis, etc). 

Of note, this will allow a common shared view of pub keys for multi-master setups.

To avoid a break-the-world scenario in this change, a fully backward compatible salt.cache.localfs_key_backcompat is provided as the new default that emulates the disk operations salt.key was doing before.

I ended up implementing master/side priv/pub key handling as well; which should cover clustered mode in totality. i think the only thing left for cluster mode to not need a glusterfs/nfs thing is the 'drop file' aes refresh mechanism which can be handled in a different PR 
    
### Tests written?
added unit tests for the cache driver

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

mattp- committed:
check_compound_minions: defer _pki_minions fetch

this can be monstrous overhead based on the amount of ACL and rules
needing to be checked, combined with how many keys you have on our
master. rather than eeagerly fetch _pki_minions, we wrap it in a closure
so that it only gets fetched if the input compound expression actually
requires it. we further only pass it to the sub matchers if provided,
as these backends can further optimize their fetches if the query is on
the full pki set.

mattp- committed:
new feature: refactor server-side PKI

the end goal of this work is to support a non file/disk based interface
for pub key management. Using the existing salt.cache interface allows
us to leverage existing implementations of other storage backends (ie
mysql, postgres, redis, etc). Of note, this will allow a common shared
view of pub keys for multi-master setups.

To avoid a break-the-world scenario in this change, a fully backward
compatible salt.cache.localfs_key_backcompat is provided as the new
default that emulates the disk operations salt.key was doing before.

open discussion items:
- as is i've left master side pub/priv key as is to live in
  etc/salt/pki; it COULD be moved to the same interface if thats desired

mattp- committed:
add missing ReqClient _closing var

squelches the warning in salt.transport.base

mattp- committed:
fix error handling on prep jid failure

this wasn't propagating to localclient as expected; error stacks can't
be wrapped in enc/load envelope.

dwoz commented on salt/crypt.py:
```diff
@@ -283,12 +283,24 @@ def decrypt(self, data, algorithm=OAEP_SHA1):
-                self.key = serialization.load_pem_public_key(fp.read())
-            except ValueError as exc:
-                raise InvalidKeyError("Invalid key")
+    def __init__(self, *, path=None, key=None):
```
I'd prefer we something like this

```
class PublicKey(BaseKey):
   @classmethod
   def from_file(cls,  path):
          ...
          return cls(key)

   def __init__(self, key):
        ...
```

mattp- replied:
sure 👍 that's cleaner

mattp- commented:
@twangboy appreciate the approve but still working on this one 👍 have some unaccounted for work with the cluster mode i need to look at as well as the current failing tests (also need to add more coverage)

twangboy commented:
I added `[WIP]` to the beginning of the title of this PR. When you're done, remove the `[WIP]`.

mattp- commented:
@twangboy sorry had to remove the label, something weird is going on with the github actions; i need the tests to run but its being wonky. will keep poking at it

mattp- committed:
migrate remaining master side pki to cache

this is a continuation of the previous commit where the 'open discussion
items' have been implemented. All private keys are now managed with a
new 'master_keys' bank that contains all priv/pub keys needed for
master operation. the cache interface is re-used in gen_keys for the
minion, which is slightly confusing but not the end of the world.

twangboy commented:
Looks like there's some lint failures

mattp- committed:
remove tests/pytests/unit/crypt/test_crypt_cryptodome.py

this is sufficiently tested in tests/pytests/unit/crypt/test_crypt.py

mattp- committed:
fix tests from pki refactor

mattp- commented on salt/crypt.py:
```diff
@@ -443,60 +375,111 @@ class MasterKeys(dict):
+                pub = self.cache.fetch("master_keys", "master.pub")
+                self.cache.store("master_keys", f"{self.master_id}.pem", priv)
+                self.cache.store("master_keys", f"{self.master_id}.pub", pub)
+                self.cache.flush("master_keys", "master.pem")
```
this is the only notable 'not backwards compatible' thing; i needed to make the master 'rename' master.pub/pem to ${id}.pub/pem, because in a multimaster scenario each masters key needs to be unique, otherwise they would erase each other on startup. i create a symlink (which is ignored in localfs_keys_backcompat) but not on windows because i was seeing some strange behavior / fails in the tests

twangboy replied:
I'd like to have @dwoz look at this to make sure it's good

mattp- replied:
going to re-run CI one more time to see if symlink on windows actually works. still worth calling out this change either way

mattp- replied:
seems to work actually. strange :) ill leave it as is so symlink should be on all platforms

mattp- commented:
@dwoz  @twangboy  this should be good to go now 👍

mattp- commented on salt/crypt.py:
```diff
@@ -330,100 +327,35 @@ def decrypt(self, data):
-        )
-
-
 def get_rsa_key(path, passphrase):
```
the memoize is still here, but get_rsa_key is no longer used when loading the master keys. i dont think this should be a big deal, but calling it out

mattp- committed:
push salt.utils.minions bits into salt.key / optimize matching

most of the operations on minion id are actually operations on the key
store, which is what salt.key provides. this refactors those parts of
the minion data cache api to use a salt.key object instead.

this commit also includes a critical performance optimization to avoid
calling _pki_minions only when necessary. in particular, recursive calls
in compound matching could call pki_minions multiple times; this change
passes in minions= to allow for subset matching to apply, avoiding a
full round trip. On smaller local installs this isn't a big deal, but on
a 75k minion cluster it adds up excessively.

mattp- committed:
update towncrier template for multi-line entry

also update this PR's md with a few lines to describe the changes
within.

dwoz reviewed (changes requested):
After my first in depth look things are looking pretty good here. I am still going checkout these changes and do little bit of a deeper dive / testing on it.

mattp- committed:
review from @dwoz : rename to localfs_key

mattp- commented on salt/master.py:
```diff
@@ -2479,8 +2486,12 @@ async def publish(self, clear_load):
         jid = self._prep_jid(clear_load, extra)
-        if jid is None:
-            return {"enc": "clear", "load": {"error": "Master failed to assign jid"}}
+        if jid is None or isinstance(jid, dict):
```
took me a minute to remember what was going on here. in the original codepath, if _prep_jid returns None, this is what happens:
https://github.com/saltstack/salt/blob/master/salt/client/__init__.py#L1926-L1944
its not expecting enc/payload envelope for stacktraces, just the raw dict with error. additionally 
https://github.com/saltstack/salt/blob/master/salt/master.py#L2542
prep_jid can directly return an {"error":"foo"} dict, so we should return that if returned, otherwise return the default "Master failed to assign jid" if None is returned

ive tweaked line 2490 to if jid and "error" in jid: instead, will push with the rest of review

dwoz replied:
Okay, this is messy but no fault of yours. I will mark this as resolved.

mattp- commented on salt/crypt.py:
```diff
@@ -330,100 +327,35 @@ def decrypt(self, data):
 
 
-def verify_signature(pubkey_path, message, signature, algorithm=PKCS1v15_SHA1):
+def verify_signature(key, message, signature, algorithm=PKCS1v15_SHA1):
```
hmm i think i missed some of the master side bits related to signing that also need to convert to the cache interface; will update in a separate commit

mattp- commented on salt/crypt.py:
```diff
@@ -281,14 +270,22 @@ def decrypt(self, data, algorithm=OAEP_SHA1):
-            except ValueError as exc:
-                raise InvalidKeyError("Invalid key")
+    def __init__(self, key_bytes):
+        log.debug("salt.crypt.get_rsa_pub_key: Loading public key")
```
updated multiple places where this was in the log line

mattp- committed:
address review from @dwoz

dwoz commented on salt/crypt.py:
```diff
@@ -992,7 +1025,7 @@ def handle_signin_response(self, sign_in_payload, payload):
 
         master_pubkey_path = os.path.join(self.opts["pki_dir"], self.mpub)
-        if os.path.exists(master_pubkey_path) and not PublicKey(
+        if os.path.exists(master_pubkey_path) and not PublicKey.from_file(
```
Okay, this makes sense. In the future we can probably come up with a way to make this more obvious. I'm okay with the way it is for now.

dwoz commented on salt/cache/localfs_key.py:
```diff
@@ -0,0 +1,375 @@
+
+.. versionadded:: 3008.0
+
+The ``localfs_keys_backcompat`` is a shim driver meant to allow the salt.cache
```
Missed this name change

mattp- replied:
woops, fixed.

dwoz commented on salt/cache/localfs_key.py:
```diff
@@ -0,0 +1,375 @@
+                ("pending", "minions_pre"),
+                ("accepted", "minions"),
+            ]:
+                keyfile = Path(cachedir, bank, key)
```
Needs `clean_path` check like on line [line 180](https://github.com/saltstack/salt/pull/67799/files#diff-a824d74f49f472a38571ecbe483f18286d4b95cd05ac041928b9d475d24cb45eR180)

mattp- replied:
hm, thought that check would cover the below. let me just remove 180 and dupe to the others

dwoz commented on salt/cache/localfs_key.py:
```diff
@@ -0,0 +1,375 @@
+        elif bank == "denied_keys":
+            # there can be many denied keys per minion post refactor, but only 1
+            # with the filesystem, so return a list of 1
+            pubfn_denied = os.path.join(cachedir, "minions_denied", key)
```
Needs `clean_path` check

dwoz commented on salt/cache/localfs_key.py:
```diff
@@ -0,0 +1,375 @@
+            if key in NON_CLUSTERED_MASTER_KEYS:
+                cachedir = __opts__["pki_dir"]
+
+            keyfile = Path(cachedir, key)
```
needs `clean_path` check

mattp- replied:
fixed

dwoz commented on salt/cache/localfs_key.py:
```diff
@@ -0,0 +1,375 @@
+            if bank in ["keys", "denied_keys"] and not valid_id(__opts__, item):
+                log.error("saw invalid id %s, discarding", item)
+
+            keyfile = Path(cachedir, base, item)
```
`clean_path` check

mattp- replied:
fixed

dwoz reviewed (changes requested):
Looking better, I still need to do a deep dive on the changes in `salt/channel/server.py`

dwoz commented:
@mattp- Please rebase, master branch should have passing tests now.
